zeroeq/client.h

```diff
@@ -0,0 +1,139 @@
+
+namespace zeroeq
+{
+/**
+ * Requests a remote procedure call on a Server.
+ *
+ * If the client is in the same session as discovered servers, it
+ * automatically connects to those servers. Requests are executed round-robin on
+ * all connected servers.
+ *
+ * A connection to a non-existing server is valid. Requests will be executed
+ * once the servers are available.
+ *
+ * A receive on any instance of a shared group will work on all instance
+ * and call the registered handlers.
+ *
+ * Not thread safe.
+ *
+ * Example: @include tests/reqRep.cpp
+ */
+class Client : public Receiver
+{
+public:
+    /**
+     * Create a default client.
+     *
+     * Postconditions:
+     * - discovers servers on _zeroeq_rep._tcp ZeroConf service
+     * - filters session \<username\> or ZEROEQ_SESSION from environment
+     *
+     * @throw std::runtime_error if ZeroConf is not available
+     */
+    ZEROEQ_API Client();
+
+    /**
+     * Create a client which connects to servers from the given session.
+     *
+     * Postconditions:
+     * - discovers publishers on _zeroeq_rep._tcp ZeroConf service
+     * - filters for given session
+     *
+     * @param session session name used for filtering of discovered publishers
+     * @throw std::runtime_error if ZeroConf is not available
+     */
+    ZEROEQ_API explicit Client(const std::string& session);
+
+    /**
+     * Create a client which subscribes to specific servers.
+     *
+     * Postconditions:
+     * - connected to the servers on the given URIs once servers are running
+     *
+     * @param uris publisher URIs in the format *|host|IP|IF:port
+     * @throw std::runtime_error if URI is not fully qualified
+     */
+    ZEROEQ_API explicit Client(const URIs& uris);
+
+    /**
+     * Create a default shared client.
+     *
+     * @sa Client()
+     * @param shared another receiver to share data reception with
+     */
+    ZEROEQ_API explicit Client(Receiver& shared);
+
+    /**
+     * Create a shared client which subscribes to servers from the given
+     * session.
+     *
+     * @sa Client( const std::string& )
+     *
+     * @param session only subscribe to servers of the same session
+     * @param shared another receiver to share data reception with
+     */
+    ZEROEQ_API Client(const std::string& session, Receiver& shared);
+
+    /**
+     * Create a shared client which subscribes to servers on the given URIs.
+     *
+     * @sa Client( const URIs& )
+     *
+     * @param uri server URIs in the format *|host|IP|IF:port
+     * @param shared another receiver to share data reception with
+     */
+    ZEROEQ_API Client(const URIs& uri, Receiver& shared);
+
+    /** Destroy this client. */
+    ZEROEQ_API ~Client();
+
+    /**
+     * Request the execution of the given data on a connected Server.
+     *
+     * The reply function will be executed during receive(). May block of when
+     * all servers are overloaded or no server is connected.
```

Comment:
One can connect to a zeroconf session which will dynamically update the servers as they are available. The Monitor can be used to get notifications.